book/src/introduction.md

```diff
@@ -8,18 +8,25 @@ intended for people that are new to JNI, and would like to learn through
 programming exercises. Suggestions, corrections and improvements are welcome and
 wanted.
 
-JNI has a lot of sharp edges. Special code needs to be written so that Java can
-call native functions (which must be built for each supported platform), and
-misuse can lead to resource leaks or JVM crashes. Due to the drawbacks of JNI,
-it is often worth considering alternatives (like reimplementation in one
-language or IPC/RPC). Eventually, [Project
-Panama](https://openjdk.java.net/projects/panama/) may make it unnecessary to
-write JNI code for some applications. If you need to write JNI code, hopefully
-you can have fun doing it in Rust.
+JNI has a lot of sharp edges. Native code needs to be written for each supported
+platform so that Java can can interact with native functions, and misuse can
+lead to resource leaks or JVM crashes. Due to the drawbacks of JNI, it is often
+worth considering alternatives (like reimplementation in one language or
+IPC/RPC). Eventually, [Project
+Panama](https://openjdk.java.net/projects/panama/) will make it [unnecessary to
+write JNI-like
+bindings](https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_jextract.md)
```

Comment:
The Panama features include:
- Linking to a native library — it lands in 16 as an incubator feature and already provides an alternative to JNI. Instead of implementing native methods, users will be able to assemble some MethodHandles: 
  - https://openjdk.java.net/jeps/389
  - https://github.com/openjdk/panama-foreign/blob/foreign-jextract/doc/panama_ffi.md#appendix-full-source-code
- jextract — will assemble these MethodHandles for you based on the library header.

Reply:
Thanks :) I've updated it to include these, let me know what you think.